src/train.jl

```diff
@@ -88,32 +89,53 @@ It adds only a few featurs to the loop above:
       (This is to move away from Zygote's "implicit" parameter handling, with `Grads`.)
     * Instead of `loss` being a function which accepts only the data,
       now it must also accept the `model` itself, as the first argument.
-    * `data` must iterate tuples, otherwise you get an error.
-      (Previously non-tuple types were not splatted into the loss. 
-      Pass in `((d,) for d in data)` to simulate this.)
+    * If `data` iterates over tuples, these will be splatted when passed to `loss`.
+      If you want to avoid the splatting, you can pass `((d,) for d in data)` instead.
     * `opt` should be the result of [`Flux.setup`](@ref). Using an optimiser
       such as `Adam()` without this step should give you a warning.
     * Callback functions are not supported.
       But any code can be included in the above `for` loop.
 """
 function train!(loss, model, data, opt; cb = nothing)
-  isnothing(cb) || error("""train! does not support callback functions.
-                            For more control use a loop with `gradient` and `update!`.""")
+  cb = old_cb_deprecation(cb)
+  cb = runall(cb)
   @withprogress for (i,d) in enumerate(data)
-    d isa Tuple || error("""train! expects as data an iterator producing tuples, but got $(typeof(d)).
-                            Pass it `((d,) for d in data)`, or use `gradient` and `update!` for more control.""")
-    l, gs = Zygote.withgradient(m -> loss(m, d...), model)
+    ds = batchmemaybe(d)
+    l, gs = Zygote.withgradient(m -> loss(m, ds...), model)
```

Comment:
Clearer to directly do what this does? Any maybe the name should not echo `gs` which is sort-of plural...
```suggestion
    d_splat = d isa Tuple ? d : (d,)
    l, gs = Zygote.withgradient(m -> loss(m, d_splat...), model)
```